metals/src/main/scala/scala/meta/internal/metals/JarTopLevels.scala

```diff
@@ -0,0 +1,111 @@
+package scala.meta.internal.metals
+
+import java.nio.file.Files
+import java.nio.file.attribute.BasicFileAttributeView
+import java.sql.{Connection, PreparedStatement, Statement}
+
+import scala.collection.concurrent.TrieMap
+import scala.meta.internal.io.PlatformFileIO
+import scala.meta.internal.metals.JdbcEnrichments._
+import scala.meta.io.AbsolutePath
+
+/**
+ * Handles caching of Jar Top Level Symbols in H2
+ *
+ * Wrapper around the indexed_jar and toplevel_symbol sql tables.
+ */
+final class JarTopLevels(conn: () => Connection) {
+
+  /**
+   * Retrieves top level Scala symbols of a jar from H2
+   *
+   * @param path absolute path of the jar
+   * @param digest MD5 digest of the jar
+   * @return the top level Scala symbols in the jar
+   */
+  def getTopLevels(
+      path: AbsolutePath
+  ): Option[TrieMap[String, AbsolutePath]] = {
+    val fs = PlatformFileIO.newJarFileSystem(path, create = false)
+    conn()
+      .query(
+        "select id from indexed_jar where path=? and last_modified=? and size=?"
+      ) { q =>
+        val attributes = getAttributes(path)
+        q.setString(1, path.toString)
+        q.setLong(2, attributes.lastModifiedTime().toMillis)
+        q.setLong(3, attributes.size())
+      } { _.getInt(1) }
+      .map { jar =>
+        val toplevels = TrieMap[String, AbsolutePath]()
+        conn().query(
+          "select symbol, path from toplevel_symbol where jar=?"
```

Comment:
Is it possible to do this in a single query by using joins?

Reply:
Will do, my nosql-damaged brain didn't even consider a join

Reply:
Left outer join added, only one query now.